File: app/src/main/java/com/machiav3lli/backup/fragments/PermissionsFragment.java
/*
 * OAndBackupX: open-source apps backup and restore app.
 * Copyright (C) 2020  Antonios Hazim
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package com.machiav3lli.backup.fragments;

import android.app.Activity;
import android.content.ActivityNotFoundException;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.net.Uri;
import android.os.Bundle;
import android.os.PowerManager;
import android.provider.Settings;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.Fragment;

import com.machiav3lli.backup.Constants;
import com.machiav3lli.backup.R;
import com.machiav3lli.backup.activities.IntroActivityX;
import com.machiav3lli.backup.databinding.FragmentPermissionsBinding;
import com.machiav3lli.backup.utils.PrefUtils;

import java.util.Arrays;

import static android.content.pm.PackageManager.PERMISSION_GRANTED;

public class PermissionsFragment extends Fragment {
    private static final String TAG = Constants.classTag(".PermissionsFragment");
    private FragmentPermissionsBinding binding;
    PowerManager powerManager;
    SharedPreferences prefs;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = FragmentPermissionsBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        setupViews();
        setupOnClicks();
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        prefs = PrefUtils.getPrivateSharedPrefs(requireContext());
        powerManager = (PowerManager) requireContext().getSystemService(Context.POWER_SERVICE);
    }

    @Override
    public void onResume() {
        super.onResume();
        updateState();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }

    private void setupViews() {
        binding.cardStoragePermission.setVisibility(PrefUtils.checkStoragePermissions(requireContext()) ? View.GONE : View.VISIBLE);
        binding.cardStorageLocation.setVisibility(PrefUtils.isStorageDirSetAndOk(requireContext()) ? View.GONE : View.VISIBLE);
        binding.cardUsageAccess.setVisibility(PrefUtils.checkUsageStatsPermission(requireContext()) ? View.GONE : View.VISIBLE);
        binding.cardBatteryOptimization.setVisibility(PrefUtils.checkBatteryOptimization(requireContext(), prefs, powerManager) ? View.GONE : View.VISIBLE);
    }

    private void setupOnClicks() {
        binding.cardStoragePermission.setOnClickListener(view -> PrefUtils.getStoragePermission(requireActivity()));
        binding.cardStorageLocation.setOnClickListener(view -> PrefUtils.requireStorageLocation(this));
        binding.cardUsageAccess.setOnClickListener(view -> getUsageStatsPermission());
        binding.cardBatteryOptimization.setOnClickListener(view -> showBatteryOptimizationDialog(powerManager));
    }

    private void updateState() {
        if (PrefUtils.checkStoragePermissions(requireContext()) &&
                PrefUtils.isStorageDirSetAndOk(requireContext()) &&
                PrefUtils.checkUsageStatsPermission(requireContext()) &&
                (prefs.getBoolean(Constants.PREFS_IGNORE_BATTERY_OPTIMIZATION, false)
                        || powerManager.isIgnoringBatteryOptimizations(requireContext().getPackageName()))) {
            moveTo(3);
        } else {
            setupViews();
        }
    }

    private void moveTo(int position) {
        ((IntroActivityX) requireActivity()).moveTo(position);
    }

    private void getUsageStatsPermission() {
        new AlertDialog.Builder(requireContext())
                .setTitle(R.string.grant_usage_access_title)
                .setMessage(R.string.grant_usage_access_message)
                .setPositiveButton(R.string.dialog_approve,
                        (dialog, which) -> startActivity(new Intent(Settings.ACTION_USAGE_ACCESS_SETTINGS)))
                .setNeutralButton(getString(R.string.dialog_refuse), (dialog, which) -> {
                })
                .setCancelable(false)
                .show();
    }

    private void showBatteryOptimizationDialog(PowerManager powerManager) {
        new AlertDialog.Builder(requireContext())
                .setTitle(R.string.ignore_battery_optimization_title)
                .setMessage(R.string.ignore_battery_optimization_message)
                .setPositiveButton(R.string.dialog_approve, (dialog, which) -> {
                    Intent intent = new Intent(Settings.ACTION_REQUEST_IGNORE_BATTERY_OPTIMIZATIONS);
                    intent.setData(Uri.parse("package:" + requireContext().getPackageName()));
                    try {
                        startActivity(intent);
                        prefs.edit().putBoolean(Constants.PREFS_IGNORE_BATTERY_OPTIMIZATION, powerManager.isIgnoringBatteryOptimizations(requireContext().getPackageName())).apply();
                    } catch (ActivityNotFoundException e) {
                        Log.w(PermissionsFragment.TAG, "Ignore battery optimizations not supported", e);
                        Toast.makeText(requireContext(), R.string.ignore_battery_optimization_not_supported, Toast.LENGTH_LONG).show();
                        prefs.edit().putBoolean(Constants.PREFS_IGNORE_BATTERY_OPTIMIZATION, true).apply();
                    }
                })
                .setNeutralButton(R.string.dialog_refuse, (dialog, which) -> prefs.edit().putBoolean(Constants.PREFS_IGNORE_BATTERY_OPTIMIZATION, true).apply())
                .setCancelable(false)
                .show();
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        if (requestCode == PrefUtils.WRITE_PERMISSION) {
            if (grantResults.length > 0 && grantResults[0] == PERMISSION_GRANTED) {
                Log.w(PermissionsFragment.TAG, String.format("Permissions were granted: %s -> %s",
                        Arrays.toString(permissions), Arrays.toString(grantResults)));
                if (!PrefUtils.canAccessExternalStorage(requireContext())) {
                    Toast.makeText(requireContext(), "Permissions were granted but because of an android bug you have to restart your phone",
                            Toast.LENGTH_LONG).show();
                }
            } else {
                Log.w(PermissionsFragment.TAG, String.format("Permissions were not granted: %s -> %s",
                        Arrays.toString(permissions), Arrays.toString(grantResults)));
                Toast.makeText(requireContext(), getString(R.string.permission_not_granted), Toast.LENGTH_LONG).show();
            }
        } else {
            Log.w(PermissionsFragment.TAG, String.format("Unknown permissions request code: %s", requestCode));
        }
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        switch (requestCode) {
            case PrefUtils.BACKUP_DIR:
                Uri uri = data.getData();
                if (resultCode == Activity.RESULT_OK) {
                    PrefUtils.setStorageRootDir(this.getContext(), uri);
                }
                break;
        }
    }
}

File: app/src/main/java/com/machiav3lli/backup/fragments/PermissionsFragment.kt
/*
 * OAndBackupX: open-source apps backup and restore app.
 * Copyright (C) 2020  Antonios Hazim
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package com.machiav3lli.backup.fragments

import android.app.Activity
import android.content.ActivityNotFoundException
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Bundle
import android.os.PowerManager
import android.provider.Settings
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AlertDialog
import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Scaffold
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import androidx.fragment.app.Fragment
import com.machiav3lli.backup.PREFS_IGNORE_BATTERY_OPTIMIZATION
import com.machiav3lli.backup.R
import com.machiav3lli.backup.activities.IntroActivityX
import com.machiav3lli.backup.databinding.FragmentComposeBinding
import com.machiav3lli.backup.ui.compose.item.PermissionItem
import com.machiav3lli.backup.ui.compose.theme.AppTheme
import com.machiav3lli.backup.ui.item.Permission
import com.machiav3lli.backup.utils.WRITE_PERMISSION
import com.machiav3lli.backup.utils.canAccessExternalStorage
import com.machiav3lli.backup.utils.checkBatteryOptimization
import com.machiav3lli.backup.utils.checkCallLogsPermission
import com.machiav3lli.backup.utils.checkContactsPermission
import com.machiav3lli.backup.utils.checkSMSMMSPermission
import com.machiav3lli.backup.utils.checkUsageStatsPermission
import com.machiav3lli.backup.utils.getPrivateSharedPrefs
import com.machiav3lli.backup.utils.getStoragePermission
import com.machiav3lli.backup.utils.hasStoragePermissions
import com.machiav3lli.backup.utils.isStorageDirSetAndOk
import com.machiav3lli.backup.utils.requireCallLogsPermission
import com.machiav3lli.backup.utils.requireContactsPermission
import com.machiav3lli.backup.utils.requireSMSMMSPermission
import com.machiav3lli.backup.utils.requireStorageLocation
import com.machiav3lli.backup.utils.setBackupDir
import timber.log.Timber

class PermissionsFragment : Fragment() {
    private lateinit var binding: FragmentComposeBinding
    private lateinit var powerManager: PowerManager
    private lateinit var prefs: SharedPreferences

    private val askForDirectory =
        registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
            if (result.data != null && result.resultCode == Activity.RESULT_OK) {
                result.data?.let {
                    val uri = it.data ?: return@registerForActivityResult
                    val flags = it.flags and (Intent.FLAG_GRANT_READ_URI_PERMISSION
                            or Intent.FLAG_GRANT_WRITE_URI_PERMISSION)
                    requireContext().contentResolver.takePersistableUriPermission(uri, flags)
                    requireContext().setBackupDir(uri)
                }
            }
        }

    private val smsmmsPermission: Unit
        get() {
            AlertDialog.Builder(requireContext())
                .setTitle(R.string.smsmms_permission_title)
                .setMessage(R.string.grant_smsmms_message)
                .setPositiveButton(R.string.dialog_approve) { _: DialogInterface?, _: Int ->
                    requireActivity().requireSMSMMSPermission()
                }
                .setNeutralButton(getString(R.string.dialog_refuse)) { _: DialogInterface?, _: Int -> }
                .setCancelable(false)
                .show()
        }

    private val callLogsPermission: Unit
        get() {
            AlertDialog.Builder(requireContext())
                .setTitle(R.string.calllogs_permission_title)
                .setMessage(R.string.grant_calllogs_message)
                .setPositiveButton(R.string.dialog_approve) { _: DialogInterface?, _: Int ->
                    requireActivity().requireCallLogsPermission()
                }
                .setNeutralButton(getString(R.string.dialog_refuse)) { _: DialogInterface?, _: Int -> }
                .setCancelable(false)
                .show()
        }

    private val contactsPermission: Unit
        get() {
            AlertDialog.Builder(requireContext())
                .setTitle(R.string.contacts_permission_title)
                .setMessage(R.string.grant_contacts_message)
                .setPositiveButton(R.string.dialog_approve) { _: DialogInterface?, _: Int ->
                    requireActivity().requireContactsPermission()
                }
                .setNeutralButton(getString(R.string.dialog_refuse)) { _: DialogInterface?, _: Int -> }
                .setCancelable(false)
                .show()
        }

    private val usageStatsPermission: Unit
        get() {
            AlertDialog.Builder(requireContext())
                .setTitle(R.string.grant_usage_access_title)
                .setMessage(R.string.grant_usage_access_message)
                .setPositiveButton(R.string.dialog_approve) { _: DialogInterface?, _: Int ->
                    startActivity(Intent(Settings.ACTION_USAGE_ACCESS_SETTINGS))
                }
                .setNeutralButton(getString(R.string.dialog_refuse)) { _: DialogInterface?, _: Int -> }
                .setCancelable(false)
                .show()
        }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        super.onCreate(savedInstanceState)
        binding = FragmentComposeBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding.composeView.setContent {
            PermissionsPage()
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        prefs = requireContext().getPrivateSharedPrefs()
        powerManager = requireContext().getSystemService(Context.POWER_SERVICE) as PowerManager
    }

    override fun onResume() {
        super.onResume()
        updateState()
    }

    @OptIn(ExperimentalMaterial3Api::class)
    @Composable
    private fun PermissionsPage() {
        val permissionsList = buildList {
            if (!requireContext().hasStoragePermissions)
                add(Pair(Permission.StorageAccess) { requireActivity().getStoragePermission() })
            if (!requireContext().isStorageDirSetAndOk)
                add(Pair(Permission.StorageLocation) {
                    requireActivity().requireStorageLocation(askForDirectory)
                })
            if (!requireContext().checkBatteryOptimization(prefs, powerManager))
                add(Pair(Permission.BatteryOptimization) {
                    showBatteryOptimizationDialog(powerManager)
                })
            if (!requireContext().checkUsageStatsPermission)
                add(Pair(Permission.UsageStats) { usageStatsPermission })
            if (!requireContext().checkSMSMMSPermission)
                add(Pair(Permission.SMSMMS) { smsmmsPermission })
            if (!requireContext().checkCallLogsPermission)
                add(Pair(Permission.CallLogs) { callLogsPermission })
            if (!requireContext().checkContactsPermission)
                add(Pair(Permission.Contacts) { contactsPermission })
        }

        AppTheme(
            darkTheme = isSystemInDarkTheme()
        ) {
            Scaffold {
                LazyColumn(
                    modifier = Modifier.fillMaxWidth(),
                    verticalArrangement = Arrangement.spacedBy(8.dp),
                    contentPadding = PaddingValues(8.dp)
                ) {
                    items(permissionsList) {
                        PermissionItem(it.first, it.second)
                    }
                }
            }
        }
    }

    private fun updateState() {
        if (requireContext().hasStoragePermissions &&
            requireContext().isStorageDirSetAndOk &&
            requireContext().checkSMSMMSPermission &&
            requireContext().checkCallLogsPermission &&
            requireContext().checkContactsPermission &&
            requireContext().checkUsageStatsPermission &&
            (prefs.getBoolean(PREFS_IGNORE_BATTERY_OPTIMIZATION, false)
                    || powerManager.isIgnoringBatteryOptimizations(requireContext().packageName))
        ) {
            (requireActivity() as IntroActivityX).moveTo(3)
        } else {
            binding.composeView.setContent {
                PermissionsPage()
            }
        }
    }

    private fun showBatteryOptimizationDialog(powerManager: PowerManager?) {
        AlertDialog.Builder(requireContext())
            .setTitle(R.string.ignore_battery_optimization_title)
            .setMessage(R.string.ignore_battery_optimization_message)
            .setPositiveButton(R.string.dialog_approve) { _: DialogInterface?, _: Int ->
                val intent = Intent(Settings.ACTION_REQUEST_IGNORE_BATTERY_OPTIMIZATIONS)
                intent.data = Uri.parse("package:" + requireContext().packageName)
                try {
                    startActivity(intent)
                    prefs.edit().putBoolean(
                        PREFS_IGNORE_BATTERY_OPTIMIZATION,
                        powerManager?.isIgnoringBatteryOptimizations(requireContext().packageName) == true
                    ).apply()
                } catch (e: ActivityNotFoundException) {
                    Timber.w(e, "Ignore battery optimizations not supported")
                    Toast.makeText(
                        requireContext(),
                        R.string.ignore_battery_optimization_not_supported,
                        Toast.LENGTH_LONG
                    ).show()
                    prefs.edit().putBoolean(PREFS_IGNORE_BATTERY_OPTIMIZATION, true).apply()
                }
            }
            .setNeutralButton(R.string.dialog_refuse) { _: DialogInterface?, _: Int ->
                prefs.edit()
                    ?.putBoolean(PREFS_IGNORE_BATTERY_OPTIMIZATION, true)?.apply()
            }
            .setCancelable(false)
            .show()
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        if (requestCode == WRITE_PERMISSION) {
            if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                Timber.w("Permissions were granted: $permissions -> $grantResults")
                if (!requireContext().canAccessExternalStorage) {
                    Toast.makeText(
                        requireContext(),
                        "Permissions were granted but because of an android bug you have to restart your phone",
                        Toast.LENGTH_LONG
                    ).show()
                }
            } else {
                Timber.w("Permissions were not granted: $permissions -> $grantResults")
                Toast.makeText(
                    requireContext(),
                    getString(R.string.permission_not_granted),
                    Toast.LENGTH_LONG
                ).show()
            }
        } else {
            Timber.w("Unknown permissions request code: $requestCode")
        }
    }
}